Update UFS to develop as of 2023 Apr 17

WalterKolczynski-NOAA opened the pull request:
**Description**
Updates the UFS model hash to the version as of 2023 Apr 17.

Some associated changes accompany this update
- Restart filenames for MOM6 (ufs-community/ufs-weather-model#1599)
- Remove store coriolis setting from MOM6 namelist (ufs-community/ufs-weather-model#1599)
- Change in atm 'log' file names (ufs-community/ufs-weather-model#1704)
- Additions to diag_table for frozen species (ufs-community/ufs-weather-model#1529)
- Restart quilting (ufs-community/ufs-weather-model#1633)
- Update to post itag (ufs-community/ufs-weather-model#1690)

The switch to restart quilting adds an additional constraint on the size of write groups to be divisible by the number of tiles, so all were increased to the next multiple of 6.

In the process of updating the diag tables, unused tables were removed.

Closes #1279 as moot
Closes #1445
Closes #1499
Partially addresses #1277

**Type of change**
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Cycled test on Orion
- [x] P8 test on Orion
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

WalterKolczynski-NOAA committed:
Update UFS to develop as of 2023 Apr 17

Updates the UFS model hash to the version as of 2023 Apr 17.

Some associated changes accompany this update
- Restart filenames for MOM6 (ufs-community/ufs-weather-model#1599)
- Remove store coriolis setting from MOM6 namelist (ufs-community/ufs-weather-model#1599)
- Change in atm 'log' file names (ufs-community/ufs-weather-model#1704)
- Additions to diag_table for frozen species (ufs-community/ufs-weather-model#1529)
- Restart quilting (ufs-community/ufs-weather-model#1633)
- Update to post itag (ufs-community/ufs-weather-model#1690)

The switch to restart quilting adds an additional constraint on the size of
write groups to be divisible by the number of tiles, so all were increased
to the next multiple of 6.

In the process of updating the diag tables, unused tables were removed.

Closes #1445
Closes #1499
Partially addresses #1277
See also #1500

WalterKolczynski-NOAA committed:
Update UFS to develop as of 2023 Apr 17

Updates the UFS model hash to the version as of 2023 Apr 17.

Some associated changes accompany this update
- Restart filenames for MOM6 (ufs-community/ufs-weather-model#1599)
- Remove store coriolis setting from MOM6 namelist (ufs-community/ufs-weather-model#1599)
- Change in atm 'log' file names (ufs-community/ufs-weather-model#1704)
- Additions to diag_table for frozen species (ufs-community/ufs-weather-model#1529)
- Restart quilting (ufs-community/ufs-weather-model#1633)
- Update to post itag (ufs-community/ufs-weather-model#1690)

The switch to restart quilting adds an additional constraint on the size of
write groups to be divisible by the number of tiles, so all were increased
to the next multiple of 6.

In the process of updating the diag tables, unused tables were removed.

Closes #1445
Closes #1499
Partially addresses #1277
See also #1500

JessicaMeixner-NOAA commented:
Thanks for the PR @WalterKolczynski-NOAA ! I'll review in the morning!

aerorahul commented on parm/parm_fv3diag/diag_table_cpl:
I don't think this `diag_table_cpl` is used anymore.
I suggest we delete it as well.

WalterKolczynski-NOAA replied:
Done

aerorahul commented on parm/parm_fv3diag/diag_table_da:
```diff
@@ -34,6 +34,13 @@
 "gfs_dyn",     "hs",          "hgtsfc",       "fv3_history",    "all",  .false.,  "none",  2
 
 "gfs_phys",    "cldfra",        "cldfra",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frzr",          "frzr",          "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frzrb",         "frzrb",         "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frozr",         "frozr",         "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frozrb",        "frozrb",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "tsnowp",        "tsnowp",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "tsnowpb",       "tsnowpb",       "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "rhonewsn",      "rhonewsn",      "fv3_history2d",  "all",  .false.,  "none",  2
```
Does the GSI read/need these? 
@CatherineThomas-NOAA @emilyhcliu 
I suggest we don't increase the volume of our output unless this is information needed by the DA executables.

WalterKolczynski-NOAA replied:
I wasn't sure myself, but forgot I wanted to ask DA about them when I opened the PR. Thanks Rahul.

CatherineThomas-NOAA replied:
No, the GSI only reads the 3D hydrometeors.  

@jiaruidong2017 @CoryMartin-NOAA - Does the snow DA use these 2D variables at all?

CoryMartin-NOAA replied:
snow DA is currently using FMS restarts. These are the sfcfNNN history files, right? If so, snow DA is unaffected by this table.

aerorahul replied:
Thanks @CoryMartin-NOAA @CatherineThomas-NOAA 
We will remove adding these fields for now.
Please open an issue if you wish to add these back in to the surface Gaussian files for the GDAS cycle.

WalterKolczynski-NOAA replied:
Removed. Thanks everyone.

aerorahul commented on parm/config/config.resources:
```diff
@@ -522,8 +522,7 @@ elif [[ "${step}" = "fcst" || "${step}" = "efcs" ]]; then
           (( QUILTPETS = ntasks_quilt * nthreads_fv3 ))
-          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP * nthreads_fv3 ))  # when threads are used, WRTTASKS_PER_GROUP = INCOMING_WRTTASKS_PER_GROUP * threads
-          # model_configure should be updated to reflect the number of threads used by calling this variable write_pets_per_group
+          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP ))
```
This is incorrect.
The `WRTTASKS_PER_GROUP` need to account the threads the atmosphere is running with.

WalterKolczynski-NOAA replied:
Not anymore, apparently. I only made this change after the forecast failed due to a PE mismatch. This line is rather superfluous now though, I should've just deleted it.

WalterKolczynski-NOAA replied:
Maybe @junwang-noaa can clarify?

junwang-noaa replied:
What I see here is that the WRTTASKS_PER_GROUP is the multiple of the MPI write tasks per group * nthreads_fv3. In other words, inside the write grid comp, the number of MPI tasks is actually WRTTASKS_PER_GROUP/nthreads_fv3, which I think is the original WRTTASKS_PER_GROUP.

aerorahul replied:
I suggest just delete this line.  It is superfluous as you note.
There is some major confusion regarding threading in the write task calculation.
```suggestion
```

aerorahul reviewed (dismissed):
Please see the inline comments.

WalterKolczynski-NOAA committed:
Remove unused diag table

WalterKolczynski-NOAA committed:
Remove unused diag table

WalterKolczynski-NOAA commented:
The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?

WalterKolczynski-NOAA commented:
> The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?

Created a new issue for this: #1513

WenMeng-NOAA commented:
> > The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?
> 
> Created a new issue for this: #1513

The UPP read 3D "refl_10cm" from model output gfs*sfcf*. @WalterKolczynski-NOAA

WalterKolczynski-NOAA committed:
Undo change to DA diag table

DA does not need the 2D frozen fields.

Refs: #

WalterKolczynski-NOAA committed:
Undo change to DA diag table

DA does not need the 2D frozen fields.

Refs: #

aerorahul committed:
patch to PR1509 related to the write tasks per group update

WenMeng-NOAA reviewed (approved):
The updates for the post process part looks good for me.

aerorahul committed:
Merge branch 'develop' into patch/1509

aerorahul committed:
make it clear write tasks are per thread and per tile

WalterKolczynski-NOAA committed:
Merge branch 'feature/ufs_update' into patch/1509

WalterKolczynski-NOAA committed:
Merge pull request #25 from aerorahul/patch/1509

patch to PR1509 related to the write tasks per group update

WalterKolczynski-NOAA committed:
Define WRTTASK_PER_GROUP_PER_THREAD

JessicaMeixner-NOAA commented:
To get my C768 HR1 like forecast job to run, I needed to double the number of WRTTASK_PER_GROUP_PER_THREAD_PER_TILE 
(there's perhaps a smaller amount to run on hera, plus we still need netcdf_parallel -> netcdf unless that error was solved and I just changed this debugging the other).  

Then I also needed this change: 
```
diff --git a/parm/config/config.resources b/parm/config/config.resources
index cf010f58..1e5b7479 100644
--- a/parm/config/config.resources
+++ b/parm/config/config.resources
@@ -530,7 +530,7 @@ elif [[ "${step}" = "fcst" || "${step}" = "efcs" ]]; then
         # PETS for quilting
         if [[ "${QUILTING:-}" = ".true." ]]; then
           (( QUILTPETS = ntasks_quilt * nthreads_fv3 ))
-          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP_PER_THREAD * nthreads_fv3 ))
+          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP_PER_THREAD ))
           export WRTTASK_PER_GROUP
         else
           QUILTPETS=0
```


This is not necessarily related to this PR, but if you change the number of threads for ice, we need this (let me know I can make a separate issue/PR for this if desired):

```
diff --git a/ush/parsing_namelists_CICE.sh b/ush/parsing_namelists_CICE.sh
index cf5b0fbf..6fc8b143 100755
--- a/ush/parsing_namelists_CICE.sh
+++ b/ush/parsing_namelists_CICE.sh
@@ -25,8 +25,8 @@ fi
 # Get correct MPI options for NPROC and grid
 local processor_shape=${cice6_processor_shape:-'slenderX2'}
 local shape=${processor_shape#${processor_shape%?}}
-local NPX=$(( ICEPETS / shape )) #number of processors in x direction
-local NPY=$(( ICEPETS / NPX ))   #number of processors in y direction
+local NPX=$(( ntasks_cice6 / shape )) #number of processors in x direction
+local NPY=$(( ntasks_cice6 / NPX ))   #number of processors in y direction
 if (( $(( NX_GLB % NPX )) == 0 )); then
   local block_size_x=$(( NX_GLB / NPX ))
 else
@@ -201,7 +201,7 @@ cat > ice_in <<eof
 /
 
 &domain_nml
-   nprocs = ${ICEPETS}
+   nprocs            = ${ntasks_cice6}
    nx_global         = ${NX_GLB}
```

JessicaMeixner-NOAA commented:
This isn't from this PR, but I ran into a bug testing this PR in which line [182 of exglobal_archive.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/scripts/exglobal_archive.sh#L182) should be updated from: 
if [ "${DO_WAVE}" = "YES" ] && [ "${WAVE_RUN}" != "gdas" ]; then
to 
if [ "${DO_WAVE}" = "YES" ] ; then

as we've already checked $RUN is gfs before, so we shouldn't need that check here and WAVE_RUN is not defined and causes an error in the archive.  Testing this fix right now.

JessicaMeixner-NOAA reviewed (changes requested):
A few bug fixes are needed, but thank you so much for making this PR @WalterKolczynski-NOAA

JessicaMeixner-NOAA commented:
> This isn't from this PR, but I ran into a bug testing this PR in which line [182 of exglobal_archive.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/scripts/exglobal_archive.sh#L182) should be updated from: if [ "${DO_WAVE}" = "YES" ] && [ "${WAVE_RUN}" != "gdas" ]; then to if [ "${DO_WAVE}" = "YES" ] ; then
> 
> as we've already checked $RUN is gfs before, so we shouldn't need that check here and WAVE_RUN is not defined and causes an error in the archive. Testing this fix right now.

This fix works.  Please let me know if you'd like this fix in a new PR, happy to submit it separately if that's preferred as it's not really related to this PR.

JessicaMeixner-NOAA commented:
@WalterKolczynski-NOAA Bing let me know that all the new output from the diag_table updates look good.  Please let me know if you want me to push any of these bug fixes that are not really related to this PR but found during it's testing to a new PR or to your branch for this PR, etc.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA Bing let me know that all the new output from the diag_table updates look good. Please let me know if you want me to push any of these bug fixes that are not really related to this PR but found during it's testing to a new PR or to your branch for this PR, etc.

Just open new issues and I'll handle them on Monday (off today).

JessicaMeixner-NOAA commented:
Thanks @WalterKolczynski-NOAA I created #1548 and #1549.  Line 533 in config.resources in this PR still needs to be fixed before merging I believe.

WalterKolczynski-NOAA committed:
Change write tasks back to per thread

After additional investigation, it was found that `WRTTASK_PER_GROUP`
really has changed to a per-thread value.

Refs: #1446, #1499

JessicaMeixner-NOAA reviewed (approved):
Thanks for fixing Line 533.

aerorahul reviewed (approved):
I don't know what the updates in the ufs-weather-model are that the `model_configure` needs.  But since these edits make it work, I am ok with it.

junwang-noaa commented on parm/config/config.ufs:
```diff
@@ -197,23 +197,28 @@ case "${fv3_res}" in
+        export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE=10  # TODO: refine these numbers when a case is available
         export WRITE_GROUP_GFS=4
-        export WRTTASK_PER_GROUP_GFS=64  # TODO: refine these numbers when a case is available
+        export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE_GFS=10  # TODO: refine these numbers when a case is available
```
Can we change this to a multiple of 6, e,g, 12? It gives developers option to try writing restart files on write grid comp?

aerorahul replied:
As seen from the variable name, it is per tile. So, it gets multiplied by 6 before ending up in `model_configure`.

WalterKolczynski-NOAA replied:
@junwang-noaa The meaning of the variable has already been changed to be per tile (we multiply by 6 later), so it is impossible for the end value to not be divisible by 6.

junwang-noaa replied:
Thanks for the explanation.
